File: Modules/Dementia/Source/Math/Vector3.cpp
#include "Vector3.h"

const Vector3 Vector3::Right = Vector3(1.f, 0.f, 0.f);

const Vector3 Vector3::Front = Vector3(0.f, 0.f, 1.f);

const Vector3 Vector3::Up = Vector3(0.f, 1.f, 0.f);
//
//inline Vector3 operator* (const Vector3& v, float s)
//{
//	return Vector3(v.x * s, v.y * s, v.z * s);
//}
//
//inline Vector3 operator/ (const Vector3& v, float s)
//{
//	s = 1.f / s;
//	return Vector3(v.x * s, v.y * s, v.z * s);
//}
//
//inline Vector3 operator- (const Vector3& v)
//{
//	return Vector3(-v.x, -v.y, -v.z);
//}
//
//inline Vector3 operator+ (const Vector3& a, const Vector3& b)
//{
//	return Vector3(a.x + b.x, a.y + b.y, a.z + b.z);
//}
//
//inline Vector3 operator- (const Vector3& a, const Vector3& b)
//{
//	return Vector3(a.x - b.x, a.y - b.y, a.z - b.z);
//}

File: Modules/ScriptCore/Source/Math/Vector3.cs
using System;

public struct Vector3
{
    public float x, y, z;

    public Vector3( float x, float y, float z )
    {
        this.x = x;
        this.y = y;
        this.z = z;
    }

    public Vector3( Vector3 other )
    {
        x = other.x;
        y = other.y;
        z = other.z;
    }

    public float Length()
    {
        return (float)Math.Sqrt( (double)( x * x + y * y + z * z ) );
    }

    public static Vector3 operator +(Vector3 thiss, Vector3 other)
    {
        return new Vector3(thiss.x + other.x, thiss.y + other.y, thiss.z + other.z);
    }

    public static Vector3 operator -(Vector3 thiss, Vector3 other)
    {
        return new Vector3(thiss.x - other.x, thiss.y - other.y, thiss.z - other.z);
    }

    public static Vector3 operator *(Vector3 thiss, float other)
    {
        return new Vector3(thiss.x * other, thiss.y * other, thiss.z * other);
    }

    public override string ToString()
    {
        return $"{{{Convert.ToString(x)}, {Convert.ToString(y)}, {Convert.ToString(z)}}}";
    }
}